// ═══════════════════════════════════════════════════════════════════════════════
// NEXUS PRIME — Sovereign AGI Pulse Protocol
// The Core Constitution: Defines the heartbeat of Multi-Agent Swarm Intelligence
// ═══════════════════════════════════════════════════════════════════════════════
// Author : Meta-Orchestrator Engineering
// Version: 1.0.0
// Proto  : proto3
// ═══════════════════════════════════════════════════════════════════════════════

syntax = "proto3";

package nexus.prime;

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";
import "google/protobuf/empty.proto";

// ─────────────────────────────────────────────────────────────────────────────
// ENUMS — State Machines & Classification
// ─────────────────────────────────────────────────────────────────────────────

enum AgentStatus {
  AGENT_STATUS_UNSPECIFIED = 0;
  AGENT_STATUS_IDLE        = 1;
  AGENT_STATUS_BUSY        = 2;
  AGENT_STATUS_ERROR       = 3;
  AGENT_STATUS_OFFLINE     = 4;
  AGENT_STATUS_BOOTING     = 5;
  AGENT_STATUS_DRAINING    = 6;  // graceful shutdown
}

enum PulseType {
  PULSE_TYPE_UNSPECIFIED    = 0;
  PULSE_TYPE_HEARTBEAT     = 1;  // periodic keep-alive
  PULSE_TYPE_STATE_UPDATE  = 2;  // agent state change
  PULSE_TYPE_TASK_RESULT   = 3;  // completed task response
  PULSE_TYPE_ERROR         = 4;  // error report
  PULSE_TYPE_REGISTRATION  = 5;  // initial handshake
  PULSE_TYPE_INTENT        = 6;  // agent requesting work or cooperation
}

enum DirectiveType {
  DIRECTIVE_TYPE_UNSPECIFIED  = 0;
  DIRECTIVE_TYPE_EXECUTE_TASK = 1;  // run a command
  DIRECTIVE_TYPE_RECONFIGURE  = 2;  // change agent config
  DIRECTIVE_TYPE_SCALE        = 3;  // spin up/down resources
  DIRECTIVE_TYPE_SHUTDOWN     = 4;  // graceful termination
  DIRECTIVE_TYPE_ACK          = 5;  // acknowledgment pulse
  DIRECTIVE_TYPE_BROADCAST    = 6;  // message to all agents
}

enum TaskStatus {
  TASK_STATUS_UNSPECIFIED = 0;
  TASK_STATUS_QUEUED      = 1;
  TASK_STATUS_DISPATCHED  = 2;
  TASK_STATUS_RUNNING     = 3;
  TASK_STATUS_SUCCESS     = 4;
  TASK_STATUS_FAILED      = 5;
  TASK_STATUS_PARTIAL     = 6;
  TASK_STATUS_TIMEOUT     = 7;
}

enum AgentType {
  AGENT_TYPE_UNSPECIFIED = 0;
  AGENT_TYPE_PLANET      = 1;  // core planet agents (SHADOW-7, X-BIO, etc.)
  AGENT_TYPE_SERVICE     = 2;  // infrastructure services
  AGENT_TYPE_HUMAN       = 3;  // human operators
  AGENT_TYPE_SWARM       = 4;  // swarm sub-agent
}

// ─────────────────────────────────────────────────────────────────────────────
// SUB-MESSAGES — Composable Building Blocks
// ─────────────────────────────────────────────────────────────────────────────

// Real-time agent telemetry
message AgentMetrics {
  double   cpu_usage_percent    = 1;
  double   memory_usage_percent = 2;
  double   request_latency_ms   = 3;
  int64    active_tasks         = 4;
  int64    completed_tasks      = 5;
  int64    failed_tasks         = 6;
  double   uptime_seconds       = 7;
  map<string, double> custom_metrics = 8;  // extensible
}

// Current agent state snapshot
message AgentState {
  AgentStatus status       = 1;
  string      current_task = 2;
  AgentMetrics metrics     = 3;
  google.protobuf.Timestamp last_pulse = 4;
}

// Agent expressing intent or requesting cooperation
message AgentIntent {
  string requested_action                = 1;
  string target_agent                    = 2;   // empty = orchestrator decides
  int32  priority                        = 3;   // 1=critical, 10=low
  google.protobuf.Struct payload         = 4;   // arbitrary JSON-compatible data
  google.protobuf.Timestamp deadline     = 5;   // optional deadline
}

// Result from a completed task
message TaskResult {
  string     command_id                  = 1;   // UUID from Cortex
  TaskStatus status                      = 2;
  google.protobuf.Struct output          = 3;   // result payload
  string     error_message               = 4;
  int64      execution_time_ms           = 5;
  string     executing_agent             = 6;
}

// Task command issued by orchestrator
message TaskCommand {
  string     command_id                  = 1;
  string     command_type                = 2;   // 'scan', 'report', 'publish', etc.
  string     origin                      = 3;   // who issued it
  google.protobuf.Struct payload         = 4;
  int32      priority                    = 5;
  google.protobuf.Timestamp deadline     = 6;
  int32      max_retries                 = 7;
  int64      timeout_ms                  = 8;   // per-task timeout
}

// Routing metadata for directive delivery
message RoutingInfo {
  string          target_agent    = 1;
  repeated string fallback_agents = 2;
  int32           max_retries     = 3;
  string          routing_rule_id = 4;
}

// ─────────────────────────────────────────────────────────────────────────────
// CORE MESSAGES — The Pulse & The Directive
// ─────────────────────────────────────────────────────────────────────────────

// Agent → Orchestrator: The continuous heartbeat of every agent
message AgentPulse {
  string    agent_id                     = 1;   // unique agent name (e.g. "SHADOW-7")
  PulseType pulse_type                   = 2;
  google.protobuf.Timestamp timestamp    = 3;

  // Contextual payloads (set based on pulse_type)
  AgentState  state                      = 4;   // for HEARTBEAT, STATE_UPDATE
  AgentIntent intent                     = 5;   // for INTENT
  TaskResult  result                     = 6;   // for TASK_RESULT

  // Registration fields (for REGISTRATION pulse_type)
  repeated string capabilities           = 7;   // ["scan","report","publish",...]
  string          endpoint               = 8;   // http://service:port
  AgentType       agent_type             = 9;
  string          display_name           = 10;

  // Error fields (for ERROR pulse_type)
  string error_code                      = 11;
  string error_message                   = 12;
  string error_stack_trace               = 13;

  // Sequence tracking
  int64 sequence_number                  = 14;  // monotonic per-agent
  string correlation_id                  = 15;  // trace ID for request chains
}

// Orchestrator → Agent: Commands & directives pushed to agents
message OrchestratorDirective {
  string        directive_id             = 1;   // UUID
  DirectiveType directive_type           = 2;
  google.protobuf.Timestamp timestamp    = 3;

  // Task assignment
  TaskCommand   command                  = 4;   // for EXECUTE_TASK
  RoutingInfo   routing                  = 5;

  // System-wide state snapshot
  SystemSnapshot system_state            = 6;

  // Reconfiguration payload
  google.protobuf.Struct config          = 7;   // for RECONFIGURE

  // Acknowledgment details
  string ack_for_pulse_id                = 8;   // which pulse this acks
  string message                         = 9;   // human-readable status message

  // Sequence
  int64  sequence_number                 = 10;
  string correlation_id                  = 11;
}

// ─────────────────────────────────────────────────────────────────────────────
// REGISTRATION — One-shot agent setup
// ─────────────────────────────────────────────────────────────────────────────

message AgentRegistration {
  string          name                   = 1;
  string          display_name           = 2;
  AgentType       agent_type             = 3;
  repeated string capabilities           = 4;
  string          endpoint               = 5;
  map<string, string> metadata           = 6;
}

message RegistrationAck {
  bool   success                         = 1;
  string agent_id                        = 2;
  string message                         = 3;
  google.protobuf.Timestamp registered_at = 4;
  SystemSnapshot system_state            = 5;  // initial system view
}

// ─────────────────────────────────────────────────────────────────────────────
// COMMAND SUBMISSION — One-shot command dispatch
// ─────────────────────────────────────────────────────────────────────────────

message CommandAck {
  bool   success                         = 1;
  string command_id                      = 2;
  string target_agent                    = 3;
  TaskStatus status                      = 4;
  string message                         = 5;
}

// ─────────────────────────────────────────────────────────────────────────────
// SYSTEM STATUS — Global cluster view
// ─────────────────────────────────────────────────────────────────────────────

message SystemSnapshot {
  int32  online_agents                   = 1;
  int32  total_agents                    = 2;
  int32  queued_commands                 = 3;
  int32  running_commands                = 4;
  double cluster_cpu_percent             = 5;
  double cluster_memory_percent          = 6;
  google.protobuf.Timestamp timestamp    = 7;
  repeated AgentSummary agents           = 8;
}

message AgentSummary {
  string      name                       = 1;
  string      display_name               = 2;
  AgentStatus status                     = 3;
  AgentType   agent_type                 = 4;
  google.protobuf.Timestamp last_seen    = 5;
  string      current_task               = 6;
  repeated string capabilities           = 7;
}

// ─────────────────────────────────────────────────────────────────────────────
// SERVICE DEFINITION — The Neural Backbone
// ─────────────────────────────────────────────────────────────────────────────

service NexusPulseService {
  // ═══ CORE: Bidirectional Pulse Stream ═══
  // The heartbeat of the swarm. Agents open a permanent stream,
  // sending pulses (heartbeats, results, errors) and receiving
  // directives (tasks, reconfigurations, broadcasts) in real-time.
  rpc Pulse(stream AgentPulse) returns (stream OrchestratorDirective);

  // ═══ REGISTRATION: One-shot agent handshake ═══
  // Agent registers capabilities and receives initial system state.
  rpc RegisterAgent(AgentRegistration) returns (RegistrationAck);

  // ═══ COMMAND: Submit a task for routing ═══
  // Any agent or external system submits a command. Orchestrator
  // routes it via rules or broadcasts to the swarm.
  rpc SubmitCommand(TaskCommand) returns (CommandAck);

  // ═══ STATUS: System health snapshot ═══
  // Returns the current state of the entire NEXUS PRIME cluster.
  rpc GetSystemStatus(google.protobuf.Empty) returns (SystemSnapshot);
}
